tracing/ops_tracing/buffer.py

```diff
@@ -0,0 +1,298 @@
+        self.ids = set()
+        self._set_db_schema()
+
+    @retry
+    def _set_db_schema(self):
+        # NOTE: measure the cost of this vs two-level approach:
+        # - check table and index in read-only mode
+        # - if needed, update the DSL
+        # NOTE: ops storage sets u+rw, go-rw permissions
+        # should we follow suit?
+        with self.tx(timeout=LONG_DB_TIMEOUT) as conn:
+            conn.execute(
+                """
+                CREATE TABLE IF NOT EXISTS tracing (
+                    -- effectively auto-incremented
+                    id INTEGER PRIMARY KEY,
+                    -- observed events are more important
+                    priority INTEGER NOT NULL,
+                    -- a chunk of tracing data
+                    data BLOB NOT NULL,
+                    -- MIME type for these data
+                    mime TEXT NOT NULL
+                )
+                """
+            )
+            conn.execute(
+                """
+                CREATE INDEX IF NOT EXISTS tracing_priority_id
+                ON tracing
+                (priority, id)
+                """
+            )
+            conn.execute(
+                """
+                CREATE TABLE IF NOT EXISTS settings (
+                    -- setting key, like "url" and "ca"
+                    key TEXT PRIMARY KEY,
+                    -- the value, with None coerced to empty string
+                    value TEXT NOT NULL
+                )
+                """
+            )
+
+    @contextlib.contextmanager
+    def tx(self, *, timeout: float = DB_TIMEOUT, readonly: bool = False):
+        """Thread-safe transaction context manager."""
+        with sqlite3.connect(self.path, isolation_level=None, timeout=timeout) as conn:
+            mode = 'DEFERRED' if readonly else 'IMMEDIATE'
+            conn.execute(f'BEGIN {mode}')
+            try:
+                yield conn
+            except:
+                conn.execute('ROLLBACK')
+                raise
+            else:
+                conn.execute('COMMIT')
+
+    @retry
+    def get_destination(self) -> Config:
+        """Get the tracing destination from the database."""
+        with self.tx(readonly=True) as conn:
+            settings = {k: v for k, v in conn.execute("""SELECT key, value FROM settings""")}
+            return Config(settings.get('url') or None, settings.get('ca') or None)
+
+    @retry
+    def set_destination(self, config: Config) -> None:
+        """Update the tracing destination in the database."""
+        with self.tx() as conn:
+            conn.execute(
+                """REPLACE INTO settings(key, value) VALUES('url', ?)""", (config.url or '',)
+            )
+            conn.execute(
+                """REPLACE INTO settings(key, value) VALUES('ca', ?)""", (config.ca or '',)
+            )
+
+    @retry
+    def mark_observed(self):
+        """Mark the tracing data collected in this dispatch as higher priority."""
```

Comment:
It might help if I understood what the priority feature is useful for, but do we even need it in v1?

Reply:
Let me start an explanation doc and cover it there.
IMO, yes we do need it.